### DISCUSSION TITLE
Pattern for fetching a list and mutating a single item within the list

### DISCUSSION THREAD

**Discussion by matomesc** [TanStack/query#5037] [Q&A] [answered]
Hi, I'm new to react query and I'm using react query to fetch a list of posts and then displaying these posts. Each post can be liked by clicking on a button which triggers a mutation. Now in order to update the UI with the new number of likes I can call `queryClient.invalidateQueries({ queryKey: ['posts'] })` which will cause the entire list to be refetched. Is there a pattern that prevents having to refetch the entire list? This seems like a common use case but I haven't been able to figure it out.

Here is what I'm currently doing:

```tsx
import { useMutation, useQuery } from '@tanstack/react-query';
import ky from 'ky';

interface PostProps {
  post: { id: string; text: string, likes: number };
  onLike: () => void;
}

const Post: React.FC<PostProps> = ({ post: { text, likes }, onLike }) => {
  return (
    <div>
      <div>{text}</div>
      <div>{likes}</div>
      <button
        type="button"
        onClick={() => {
          onLike();
        }}
      >
        Like
      </button>
    </div>
  );
};

const Posts: React.FC = () => {
  const { data, status } = useQuery({
    queryKey: ['posts'],
    queryFn: async () => {
      return ky
        .get('/api/posts')
        .json<{ id: string; text: string; likes: number }[]>();
    },
  });
  const likeMutation = useMutation({
    mutationFn: (id: string) => {
      return ky.post(`/api/posts/${id}/like`).json();
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['posts'] });
    }
  });

  if (status === 'loading') {
    return <div>Loading...</div>;
  }

  if (status === 'error') {
    return <div>Error</div>;
  }

  return (
    <div>
      {data.map((post) => {
        return (
          <Post
            key={post.id}
            post={post}
            onLike={() => {
              likeMutation.mutate(post.id);
            }}
          />
        );
      })}
    </div>
  );
};
```

**Comment by TkDodo** [comment, marked as answer]
> which will cause the entire list to be refetched. Is there a pattern that prevents having to refetch the entire list? This seems like a common use case but I haven't been able to figure it out.

your query always fetches from `/api/posts`, so an invalidation will always refetch the entire list. That's often a good thing - items could have been added or altered by someone else in the meantime, so now you'd have the up-to-date data.

If you don't want that, have a look at [Updates from Mutation Responses](https://tanstack.com/query/v4/docs/react/guides/updates-from-mutation-responses). If your mutation returns the updated item, you can write it back to the list in `onSuccess`:

```
onSuccess: (item) => {
  queryClient.setQueryData(
    ['posts'],
    (prev) => prev?.map(post => post.id === item.id ? item : post)
  )
}
```

**Reply by matomesc** [reply]
That's exactly what I was looking for, thanks!

**Reply by radik-albaik** [reply]
That helped me too!

**Reply by tmiyasa** [reply]
What if the number of posts is very large, like around 1000? We would need to loop through all 1000 posts every time a user likes a post.

**Reply by radik-albaik** [reply]
IMO If you have not paginated, I'm afraid **yes**.

**Reply by TkDodo** [reply]
> We would need to loop through all 1000 posts

If you have 1000 posts, I'd be much more concerned about transporting that much content over the network or having to render 1000 posts in the UI at the same time rather than having a constant loop over 1k elements after a user interaction. Code that is linear is not something to worry about.

**Reply by IgorKurkov** [reply]
I just wondering, should we have a method like `updateOne` or `updateMany` in NgXs or typeORM for this situation? currently we could mutate the cache state by `setQueryData => prev....` callback, but this prev object is different for each cache data. what if we can update the item inside the cache with a `queryClient.<updateOne>` method? 

What do you think guys?

**Comment by silberGuy** [comment]
Hi!
I also encountered this situation, where I wanted to prevent a new fetch, and thought of a pattern that I would like to have some feedback about.

We can fetch the whole list once, and than instead of creating a single query for the list as a whole, we can create an array of queries for each individual item, to be used with `useQueries`. Something like the following:

```js
function usePostsQueries() {
  const { data } = useQuery({
    queryKey: ['posts'],
    queryFn: async () => ky.get('/api/posts').json()
  })
  
  const results = useQueries(
    (data || []).map(post => ({
      queryKey: ['posts', post.id],
      queryFn: async () => post
    }))
  )

  const resultsData = results.map(res => res.data)
  return useQuery({
    queryKey: ['posts', ...resultsData],
    queryFn: async () => resultsData
  })
}
```

By this approach, the list query hook is the one to assure it is up to date. Single item mutations can just invalidate using the key `['posts', post.id]`, and the list will update automatically.

**Reply by bfaulk96** [reply]
I've tried a similar approach to this, but it seems to break down on large datasets and cause some pretty significant performance problems (i.e. when you have hundreds of items and it has to have hundreds of active queries at once)
